plugins/usbdmx/ShowJockey.cpp

```diff
@@ -0,0 +1,392 @@
+ public:
+  ShowJockeyThreadedSender(LibUsbAdaptor *adaptor,
+                           libusb_device *usb_device,
+                           libusb_device_handle *usb_handle,
+                           int maxPacketSizeOut,
+                           int endpoint);
+
+ private:
+  LibUsbAdaptor* const m_adaptor;
+  int m_maxPacketSizeOut;
+  int m_endpoint;
+
+  bool TransmitBuffer(libusb_device_handle *handle,
+                      const DmxBuffer &buffer);
+  int bulkSync(libusb_device_handle *handle,
+               int endpoint,
+               int maxPacketSize,
+               unsigned char *buffer,
+               int size);
+};
+
+ShowJockeyThreadedSender::ShowJockeyThreadedSender(
+    LibUsbAdaptor *adaptor,
+    libusb_device *usb_device,
+    libusb_device_handle *usb_handle,
+    int maxPacketSizeOut,
+    int endpoint)
+    : ThreadedUsbSender(usb_device, usb_handle),
+      m_adaptor(adaptor),
+      m_maxPacketSizeOut(maxPacketSizeOut),
+      m_endpoint(endpoint) {
+}
+
+bool ShowJockeyThreadedSender::TransmitBuffer(libusb_device_handle *handle,
+                                              const DmxBuffer &buffer) {
+  if (!handle) {
+    return false;
+  }
+
+  int retVal = 0;
+  int leftWriteSize = 512;
+  uint16_t alreadyWrittenSize = 0;
+  unsigned int writeSize = 0;
+  int maxPacketSizeOut = m_maxPacketSizeOut;
+
+  unsigned char *bulkBuffer = new unsigned char[maxPacketSizeOut]();
+
+  int bulkDataSize = maxPacketSizeOut - 2;
+  unsigned int slot = 0;
+
+  while (leftWriteSize > 0) {
+    memset(bulkBuffer, 0, maxPacketSizeOut);
+    memcpy(bulkBuffer, &alreadyWrittenSize, 2);
+
+    if (leftWriteSize > bulkDataSize) {
+      writeSize = bulkDataSize;
+    } else {
+      writeSize = leftWriteSize;
+    }
+
+    buffer.GetRange(slot, bulkBuffer+2, &writeSize);
+
+    retVal = bulkSync(handle, m_endpoint, maxPacketSizeOut,
+                      bulkBuffer, writeSize+2);
+
+    if (retVal < 0) {
+      return retVal;
+    }
+
+    leftWriteSize -= writeSize;
+    alreadyWrittenSize += writeSize;
+    slot += writeSize;
+  }
+
+  if (bulkBuffer) {
+    delete[] bulkBuffer;
+    bulkBuffer = 0;
+  }
+
+  return true;
+}
+
+int ShowJockeyThreadedSender::bulkSync(libusb_device_handle *handle,
+                                       int endpoint, int maxPacketSize,
+                                       unsigned char *buffer, int size) {
+  if (!handle) {
+    return -1;
+  }
+
+  if (size <= maxPacketSize) {
+    int transferred = -1;
+    m_adaptor->BulkTransfer(handle, endpoint, buffer, size,
+                            &transferred, URB_TIMEOUT_MS);
+
+    return transferred;
+  } else {
+    int retVal = 1;
+    unsigned char *buf = buffer;
+    int nPacket = maxPacketSize;
+    int nLeft = size;
+    int transferred = -1;
+    int transferredAll = 0;
+
+    while (1) {
+      if (nLeft <= 0) {
+        break;
+      }
+
+      retVal = m_adaptor->BulkTransfer(handle, endpoint, buffer, nPacket,
```

Comment:
I think you should be able to still do this as one bulk transfer. See e.g. the ScanlimeFadecandy code.